install-validator/pkg/dir/upgrade.go

```diff
@@ -0,0 +1,63 @@
+package dir
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+
+	"github.com/sirupsen/logrus"
+
+	kubewrapper "github.com/vmware-tanzu/graph-framework-for-microservices/install-validator/pkg/k8s-utils"
+)
+
+// ApplyDir checks if there are any incompatible crds and data for them. Applies them based on force flag.
+func ApplyDir(directory string, force bool, c kubewrapper.ClientInt, cFunc compareFunc) error {
+	err := c.FetchCrds()
+	if err != nil {
+		return err
+	}
+
+	// check for incompatible and outdated models. Return if there are any incompatibilities and force != true
+	inCompatibleCRDs, outdated, err := CheckDir(directory, c, cFunc)
+	if err != nil {
+		return err
+	}
+	if len(inCompatibleCRDs) > 0 && !force {
+		textChanges := new(bytes.Buffer)
+		for _, txt := range inCompatibleCRDs {
+			textChanges.Write(txt.Bytes())
+		}
+		logrus.Warn(textChanges)
+		return errors.New("incompatible datamodel changes detected")
+	}
+
+	// check if there are any resources for incompatible crds and return if so
+	var cr []string
+	for crd := range inCompatibleCRDs {
+		res, err := c.ListResources(*c.GetCrd(crd))
+		if err != nil {
+			return err
+		}
+		if len(res) > 0 {
+			cr = append(cr, crd)
+		}
+	}
+	if len(cr) > 0 {
+		return fmt.Errorf("validation failed as objects exists in the system for the incompatible nodes: %v", cr)
+	}
+
+	//delete outdated models
+	for _, o := range outdated {
+		err = c.DeleteCrd(o)
```

Comment:
@rafal-slominskii  I took a deeper look into it. Couldn't find the below case addressed. Pls help me understand

PTAL
https://github.com/vmware-tanzu/graph-framework-for-microservices/pull/90#discussion_r1042619229

> If CR's exist for nonInstalled CRD's, then the installation should fail.
> If no CR's exits for notInstalled CRD"s, then these CRD's should be removed from the system, before upserting crd's from the new datamodel.


Looks like, we may need to check if the CR's (**instances**) exists for the nonInstalled CRD's

if CR's(instances) exists - we should fail 
If not - we can remove the CRD type from the system